Source/WebCore/editing/Editing.cpp

```diff
@@ -1356,55 +1371,119 @@ TextDirection primaryDirectionForSingleLineRange(const Position& start, const Po
     return direction;
 }
 
-SimpleRange adjustToVisuallyContiguousRange(const SimpleRange& range)
+static std::optional<SimpleRange> makeVisuallyContiguousIfNeeded(const SimpleRange& range, OptionSet<RangeEndpointsToAdjust> endpoints, SelectionExtentMovement movement)
 {
     if (range.collapsed())
-        return range;
+        return std::nullopt;
 
     auto [start, end] = positionsForRange(range);
     auto firstLineDirection = start.primaryDirection();
     RenderedPosition renderedStart { start };
     if (renderedStart.isNull() || renderedStart.lineBox().atEnd())
-        return range;
+        return std::nullopt;
 
     auto lastLineDirection = end.primaryDirection();
     RenderedPosition renderedEnd { end };
     if (renderedEnd.isNull() || renderedEnd.lineBox().atEnd())
-        return range;
+        return std::nullopt;
 
     if (renderedStart.box() == renderedEnd.box())
-        return range;
+        return std::nullopt;
 
     auto bidiLevelAtStart = renderedStart.box()->bidiLevel();
     auto bidiLevelAtEnd = renderedEnd.box()->bidiLevel();
     auto targetBidiLevelAtStart = bidiLevelAtStart;
     auto targetBidiLevelAtEnd = bidiLevelAtEnd;
+    auto directionAtStart = renderedStart.box()->direction();
+    auto directionAtEnd = renderedEnd.box()->direction();
     if (inSameLine(start, end)) {
         forEachRenderedBoxBetween(renderedStart, renderedEnd, [&](auto box) {
```

Comment:
Should we add a new method or a helper function which builds the lowest bidi-level between two RenderedPositions?

Reply:
Sounds good — created a helper function here:

```
static InlineIterator::LeafBoxIterator boxWithMinimumBidiLevelBetween(const RenderedPosition& start, const RenderedPosition& end)
{
    InlineIterator::LeafBoxIterator result;
    forEachRenderedBoxBetween(start, end, [&](auto box) {
        if (!result || box->bidiLevel() < result->bidiLevel())
            result = box;
    });
    return result;
}
```

...and used it in `primaryDirectionForSingleLineRange` and `makeVisuallyContiguousIfNeeded`. This also made me realize that `directionAtStart` and `directionAtEnd` are unused now, so I reverted that part.